File: scripts/fuzzer.rb

#
# fuzzer.rb

require 'io/console'

rows, cols = IO.console.winsize

lines = (File.readlines(ARGV[0]) rescue [ nil, nil, nil ])
fi0 = ARGV[1]

suffixes = %w[
  js rb ru java pl py c go
  sh fish vim
  txt text md mdown markdown
  htm html slim haml
  css scss
  xml toml json yaml yml conf cnf
  flo ]

fs = (
  Dir["**/*.{#{suffixes.join(',')}}"] +
  Dir['**/*/']#.collect { |e| e + '/' }
    ).sort

hi = (lines[1] || '').strip.split('|')
hishow = false

fi =
  if fi0 && fi0 != ''
    fi0
  else
    fi1 = hi[0] || ''; hi.rotate!; fi1
  end

li = (lines[2] || 0).to_i
pre = nil

  # 15      11      app/views/wma/_dac_4_approvals.slim
  # 67      24      app/views/wma/_dac_5_assignment.slim
  # 1       1       app/views/wma/_dac_mandates.slim
  #
$git = `git diff --numstat`.strip.split("\n")
  .inject({}) { |h, l|
    ll = l.split("\t")
    h[ll.pop] = ll
    h }
      # FIXME git root might not be here...
$gits = `git ls-files`.strip.split("\n")

def d_size(path)
  i = File.size(path)
  [ '', 'K', 'M', 'G', 'T' ].each do |n|
    return "#{i.to_i}#{n}" if i < 1024
    i = i / 1024.0
  end
  '-1'
end
def d_lines(path)
  "#{(File.readlines(path).size rescue -1)}l"
end
def d_diff(path)
  g = $git[path]
  g ? "+#{g[0]}-#{g[1]}" : nil
end
def d_recent(path)
  (Time.now - File.mtime(path)) < 24 * 60 * 60
end
def d_git(path)
  $gits.include?(path)
end
  #
def detail(path)
  $details ||= {}
  $details[path] ||=
    File.directory?(path) ?
    { dir: true, recent: d_recent(path), git: true } :
    { size: d_size(path), lines: d_lines(path), diff: d_diff(path),
      recent: d_recent(path), git: d_git(path) }
end

def select(pat, files)
  return files if pat == ''
  pats = pat.split('*')
  if pats.length < 2
    files.select { |f|
      f.downcase.index(pat) }
  else
    files.select { |f|
      f1 = f.downcase;
      f1.start_with?(pats[0]) && f1.index(pats[1]) }
  end
end

dcol = '92' # directory colour
fcol = '32' # filename colour
scol = '90' # slash colour
ocol = '93' # dot color
tcol = '90' # detail color
gcol = '97' # git color
rcol = '2;33' # recent color

recent = '<<'

path = nil

loop do

  fs1 = select(fi, fs)[0, rows - 1]

  print "[2J" # clear

  fs1.each_with_index do |f, i|

    d = detail(f)

    print "[#{2 + i};1H[0;0m  "
    color = i == li ? "[1;#{dcol};7m" : "[#{dcol}m"
    print color

    t = f[0, cols - 2]
    tt = t.split('/')

    dot = i == li ? '.' : "[#{ocol}m.#{color}[#{fcol}m"
    tt[-1] = "[#{fcol}m#{tt[-1].split('.').join(dot)}" unless d[:dir]

    print tt.join("[#{scol}m/#{color}")

    print "[#{scol}m/#{color}" if d[:dir]
    print "  [#{tcol}m#{d[:size]} #{d[:lines]}" if d[:size]

    if d[:diff]
      print "  [#{gcol}m#{d[:diff]}"
    elsif ! d[:git]
      print "  [#{gcol}mng"
    end

    print "  [#{rcol}m#{recent}" if d[:recent]
  end

  print "[1;#{cols}H"
  print "[0;0m"
  print pre || ''

  print "[1;1H"
  if hishow
    print fi + '  '
    print hi.join(' ')[0, cols - fi.size - 2]
    print "[1;1H"
    print "[7m"
    print fi # ;-)
    print "[0;0m"
    hishow = false
  else
    print "[0;0m"
    print fi
  end

  c = $stdin.raw { |io| io.readpartial(4) }

  if pre == ':'
    if c == 'q'
      path = ''; break
    #elsif c == ':'
    #  c = '['
    end
  elsif pre == ';'
    if c == ';'
      path = ''; break
    elsif c == 'b'
      path = '(buffers)'; break
    elsif c == 'd'
      path = '(diff)'; break
    elsif c == 'l'
      path = '(log)'; break
    elsif c == 't'
      path = File.dirname(fs1[li]) + '/'; break
    end
  end
  pre = nil

  if c == "\e" # Escape
    path = ''; break
  elsif c == "\r" || c == "\n" # Return / Enter
    path = fs1[li]; break
  elsif c == "\x7F" # Backspace
    fi = fi[0..-2]; li = 0
  elsif c == "\e[A" || c == 'k' # up
    li = li - 1; li = 0 if li < 1
  elsif c == "\e[B" || c == 'j' # down
    li = li + 1; li = fs1.length - 1 unless fs1[li]
  elsif c == 'G'
    li = fs1.length - 1
  elsif c == '@' || c == '['
    hi << fi; fi = hi[0] || ''; hi.uniq!.rotate!; hishow = true
  elsif c == "\t" || c == '/' # Tab / Slash
    if li == fs1.length - 1
      li = 0
    else
      dir = File.dirname(fs1[li])
      while cu = fs1[li]
        break if File.dirname(cu) != dir
        li = li + 1
      end
      li = fs1.length - 1 unless fs1[li]
    end
  elsif c == "\e[3~" || c == 'C' # Delete
    fi = ''; li = 0
  elsif c.length > 1
    p c; sleep 0.7
  elsif c == ' '
    li1 = li + 1
    #while li1 < fs1.length
    loop do
      if li1 >= fs1.length
        li = 0; break
      end
      f = fs1[li1]
      d = detail(f)
      if d && (d[:diff] || d[:recent])
        li = li1; break
      end
      li1 += 1
    end
  elsif c == ':' || c == ';'
    pre = c
  elsif ';'.index(c)
    # ignore
  else
    fi = fi + c.downcase; li = 0
  end

#rescue => err
#  p err
end

print "[2J" # clear

hi = ([ fi ] + hi).take(7).uniq.join('|')

File.open(ARGV[0], 'wb') { |f| f.puts([ path, hi, li.to_s ].join("\n")) }

File: scripts/fuzzer.py
# -*- coding: UTF-8 -*-

# fuzzer.py

import os, re, sys, glob, time, string, fnmatch, subprocess

W = int(sys.argv[1])

max_vimfuzz_lines = 21

pat = ''
opts = {}
keyr = re.compile('^([^:]+):(.+)$')
  #
try:
  lines = None
  with open('.vimfuzz2', 'r') as file:
    lines = file.readlines()
    pat = lines[-1].strip()
  with open('.vimfuzz2', 'w') as file:
    ls = []
    for l in lines:
      m = keyr.match(l)
      if m:
        file.write(l)
        opts[m.group(1)] = m.group(2)
      else:
        ls.append(l)
    while len(ls) > max_vimfuzz_lines:
      ls.pop(0)
    for l in ls:
      file.write(l)
except:
  ''
  #
patr = re.compile(pat, re.IGNORECASE);

exts = re.split(
  r"\s+",
  '''
    js rb ru java pl py c go
    sh fish vim
    txt text md mdown markdown
    htm html slim haml
    css scss
    xml toml json yaml yml conf cnf
    flo '''.strip())

#sortr = re.compile('^tmp')
#def sortPaths(path):
#  if sortr.search(path): return 'ZZZ/' + path
#  return path

files = filter(
  lambda p: os.path.splitext(p)[1][1:] in exts,
  glob.glob('**/*', recursive=True))
dirs = glob.glob(
  '**/*/', recursive=True)
paths = list(files) + dirs

for ign in opts.get('ignore', '').split(r'\s+'):
  paths = filter(lambda p: not(fnmatch.fnmatch(p, ign)), paths)

paths = filter(lambda p: patr.search(p), paths)
#paths = sorted(paths, key=sortPaths)
#paths = sorted(paths)

#
# file details

def exec_to_lines(cmd):
  return map(
    lambda l: l.decode().rstrip(),
    subprocess.Popen(cmd, shell=True, stdout=subprocess.PIPE).stdout)

git = {}
  #
for line in exec_to_lines("git diff --numstat"):
  ll = line.split("\t")
  git[ll.pop()] = ll
for line in exec_to_lines("git diff --cached --numstat"):
  ll = line.split("\t")
  git[ll.pop()] = ll
#gits = exec_to_lines("git ls-files")
  # FIXME git root might not be here...

def d_size(path):
  size = os.path.getsize(path)
  step_unit = 1024.0 # https://stackoverflow.com/questions/12523586
  for x in [ 'B', 'K', 'M', 'G', 'T', 'P' ]:
    if size < step_unit: return "%i%s" % (size, x)
    size /= step_unit

def d_lines(path):
  l = 0
  try:
    with open(path, 'r') as file:
      for line in file: l += 1
  except:
    l = -1
  return l

def d_diff(path):
  g = git.get(path)
  return f"+{g[0]}-{g[1]}" if g else ''

def d_mtime(path):
  return os.path.getmtime(path)

#def d_git(path):
#  return path in gits

def d_age(mtime):
  i = int(time.time() - mtime)
  if i < 60: return '%is' % i
  d = int(i / (24 * 3600)); i = i % (24 * 3600)
  h = int(i / 3600); i = i % 3600
  m = int(i / 60); i = i % 60
  r = ''
  if d > 0: r = '%s%id' % (r, d)
  if d < 7 and h > 0: r = '%s%ih' % (r, h)
  if d < 1 and m > 0: r = '%s%im' % (r, m)
  return r

details = {}
  #
def detail(path):
  d = details.get(path)
  if d:
    return d
  mt = d_mtime(path)
  age = d_age(mt)
  if os.path.isdir(path):
    dl = len(path) - 1
    d = { "dir": True, "mtime": mt, "age": age, "dirl": dl }
  else:
    dl = len(os.path.dirname(path))
    if os.path.basename(path) == path: dl = dl - 1
    d = {
      "size": d_size(path), "lines": d_lines(path), "diff": d_diff(path),
      "mtime": mt, "age": age, "dirl": dl }
  details[path] = d
  return d

def sortPaths(path):
  d = detail(path)
  return - d['mtime']

paths = sorted(paths)
#paths = sorted(paths, key=sortPaths)

mdl = 0
for path in paths:
  d = detail(path)
  mdl = max(mdl, d['dirl'])

#
# output

endr = re.compile('^(.+)[-+0-9 ]$')
def endstrip(s):
  s = s[:W-1].rstrip()
  while True:
    m = endr.match(s)
    if not(m): break
    s = m.group(1)
  return s

print(pat)
  #
for path in paths:
  d = detail(path)
  offset = ' ' * (mdl - d['dirl'])
  if d.get('dir'):
    print(endstrip(
      ' %s%s' % (
        offset, path)))
  else:
    print(endstrip(
      ' %s%s %s %iL %s%s' % (
        offset, path, d['size'], d['lines'], d['diff'], d['age'])))
print()

  # keep that in the fridge, but most of the time, the fuzzer is
  # called from a .git level
  #
## determine git root
#
#gitroot = False
#try:
#  DEV_NULL = open(os.devnull, 'w')
#  gitroot = subprocess\
#    .check_output([ 'git', 'rev-parse', '--show-toplevel' ], stderr=DEV_NULL)\
#    .decode()\
#    .strip()
#except:
#  True # not a git repo

